Source/FieldSolver/ImplicitSolvers/WarpXImplicitOps.cpp

```diff
@@ -0,0 +1,333 @@
+        amrex::ParticleReal* ux_n = pti.GetAttribs(particle_comps["ux_n"]).dataPtr();
+        amrex::ParticleReal* uy_n = pti.GetAttribs(particle_comps["uy_n"]).dataPtr();
+        amrex::ParticleReal* uz_n = pti.GetAttribs(particle_comps["uz_n"]).dataPtr();
+
+        const long np = pti.numParticles();
+
+        amrex::ParallelFor( np, [=] AMREX_GPU_DEVICE (long ip)
+        {
+            amrex::ParticleReal xp, yp, zp;
+            getPosition(ip, xp, yp, zp);
+
+#if (AMREX_SPACEDIM >= 2)
+            xp = 2._rt*xp - x_n[ip];
+#endif
+#if defined(WARPX_DIM_3D) || defined(WARPX_DIM_RZ)
+            yp = 2._rt*yp - y_n[ip];
+#endif
+            zp = 2._rt*zp - z_n[ip];
+
+            ux[ip] = 2._rt*ux[ip] - ux_n[ip];
+            uy[ip] = 2._rt*uy[ip] - uy_n[ip];
+            uz[ip] = 2._rt*uz[ip] - uz_n[ip];
+
+            setPosition(ip, xp, yp, zp);
+        });
+
+    }
+    }
+
+    }
+
+}
+
+void
+WarpX::FinishImplicitField( amrex::Vector<std::array< std::unique_ptr<amrex::MultiFab>, 3 > >& Field_fp,
+                      const amrex::Vector<std::array< std::unique_ptr<amrex::MultiFab>, 3 > >& Field_n,
+                      amrex::Real theta )
+{
+    using namespace amrex::literals;
+
+    for (int lev = 0; lev <= finest_level; ++lev) {
+
+#ifdef AMREX_USE_OMP
+#pragma omp parallel if (amrex::Gpu::notInLaunchRegion())
+#endif
+       for ( amrex::MFIter mfi(*Field_fp[lev][0], amrex::TilingIfNotGPU()); mfi.isValid(); ++mfi )
+        {
+
+            amrex::Array4<amrex::Real> const& Fx = Field_fp[lev][0]->array(mfi);
+            amrex::Array4<amrex::Real> const& Fy = Field_fp[lev][1]->array(mfi);
+            amrex::Array4<amrex::Real> const& Fz = Field_fp[lev][2]->array(mfi);
+
+            amrex::Array4<amrex::Real> const& Fx_n = Field_n[lev][0]->array(mfi);
+            amrex::Array4<amrex::Real> const& Fy_n = Field_n[lev][1]->array(mfi);
+            amrex::Array4<amrex::Real> const& Fz_n = Field_n[lev][2]->array(mfi);
+
+            amrex::Box const& tbx = mfi.tilebox(Field_n[lev][0]->ixType().toIntVect());
+            amrex::Box const& tby = mfi.tilebox(Field_n[lev][1]->ixType().toIntVect());
+            amrex::Box const& tbz = mfi.tilebox(Field_n[lev][2]->ixType().toIntVect());
+
+            const amrex::Real c0 = 1._rt/theta;
+            const amrex::Real c1 = 1._rt - c0;
+
+            amrex::ParallelFor(
+            tbx, ncomps, [=] AMREX_GPU_DEVICE (int i, int j, int k, int n)
+            {
+                Fx(i,j,k,n) = c0*Fx(i,j,k,n) + c1*Fx_n(i,j,k,n);
+            },
+            tby, ncomps, [=] AMREX_GPU_DEVICE (int i, int j, int k, int n)
+            {
+                Fy(i,j,k,n) = c0*Fy(i,j,k,n) + c1*Fy_n(i,j,k,n);
+            },
+            tbz, ncomps, [=] AMREX_GPU_DEVICE (int i, int j, int k, int n)
+            {
+                Fz(i,j,k,n) = c0*Fz(i,j,k,n) + c1*Fz_n(i,j,k,n);
+            });
+        }
+    }
+}
+
+void
+WarpX::ImplicitComputeRHSE (amrex::Real a_dt, WarpXSolverVec&  a_Erhs_vec)
```

Comment:
The name "compute RHS" is maybe a bit misleading. 
My understanding is that this actually "updates" the E field by applying the RHS, i.e. it performs:
```
E = E_n + RHS
```
Would `ImplicitApplyRHSE` be a more accurate name?

Reply:
As we discussed earlier, it does compute the RHS.